src/main/utility/interval_map.rs

```diff
@@ -0,0 +1,610 @@
+                let old = Interval::new(self.begins[i], self.ends[i]);
+                self.ends[i] = begin - 1;
+                mutations.push(Mutation::ModifiedEnd(old, self.ends[i]));
+            }
+        }
+
+        // Find the end of the splice interval.
+        // TODO: Maybe do binary search here.
+        while splice_end < self.ends.len() && self.ends[splice_end] <= end {
+            splice_end += 1
+        }
+
+        // Check whether we need to clip the beginning splice_end's interval.
+        let mut modified_begin: Option<Mutation<V>> = None;
+        if splice_end < self.begins.len()
+            && self.begins[splice_end] <= end
+            && self.ends[splice_end] > end
+        {
+            let i = splice_end;
+            let old = Interval::new(self.begins[i], self.ends[i]);
+            self.begins[i] = end + 1;
+            modified_begin = Some(Mutation::ModifiedBegin(old, self.begins[i]));
+        }
+
+        // Do the splice
+        let dropped_begins: Vec<_> = self
+            .begins
+            .splice(splice_start..splice_end, begins_insertions)
+            .collect();
+        let dropped_ends: Vec<_> = self
+            .ends
+            .splice(splice_start..splice_end, ends_insertions)
+            .collect();
+        {
+            // We use the dropped_vals iterator directly here to avoid extra copies.
+            // This is in a new scope to limit the lifetime of the mutable borrow from self.vals.
+            let mut dropped_vals = self.vals.splice(splice_start..splice_end, vals_insertions);
+            for i in 0..dropped_begins.len() {
+                mutations.push(Mutation::Removed(
+                    Interval::new(dropped_begins[i], dropped_ends[i]),
+                    dropped_vals.next().unwrap(),
+                ));
+            }
+        }
+
+        // Do the modified beginning, if any, last, so that mutations are ordered.
+        if let Some(m) = modified_begin {
+            mutations.push(m)
+        }
+
+        mutations
+    }
+
+    // Returns the item at the given index.
+    fn item_at(&self, i: usize) -> (Interval, &V) {
+        (Interval::new(self.begins[i], self.ends[i]), &self.vals[i])
+    }
+
+    // Returns the index of the interval containing `x`.
+    fn get_index(&self, x: usize) -> Option<usize> {
+        match self.begins.binary_search(&x) {
+            Ok(i) => Some(i),
+            Err(i) => {
+                if i == 0 {
+                    None
+                } else if x <= self.ends[i - 1] {
+                    Some(i - 1)
+                } else {
+                    None
+                }
+            }
+        }
+    }
+
+    // Returns the entry of the interval containing `x`.
+    pub fn get(&self, x: usize) -> Option<(Interval, &V)> {
+        match self.get_index(x) {
+            None => None,
+            Some(i) => Some(self.item_at(i)),
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
```

Comment:
There's one other test I think would be useful (unless I missed it). One where it adds an interval with a begin that matches a begin from another interval. Something like `m.insert(1, 5, "i1".to_string())`, and then `m.insert(1, 3, "i2".to_string())`. This would make sure the indexing works properly when the binary search returns the exact `splice_start`.

Reply:
Good idea - done